querybook/webapp/components/QueryComposer/RunQuery.tsx

```diff
@@ -71,14 +75,58 @@ async function transformTemplatedQuery(
     }
 }
 
-function transformLimitedQuery(
+async function transformLimitedQuery(
     query: string,
     rowLimit: Nullable<number>,
     engine: IQueryEngine
 ) {
-    return engine.feature_params?.row_limit && rowLimit != null
-        ? getLimitedQuery(query, rowLimit, engine.language)
-        : query;
+    if (!engine.feature_params?.row_limit) {
+        return query;
+    }
+
+    if (rowLimit != null && rowLimit >= 0) {
+        return getLimitedQuery(query, rowLimit, engine.language);
+    }
+
+    // query is unlimited but engine has row limit feature turned on
+
+    const unlimitedSelectQuery = hasQueryContainUnlimitedSelect(
+        query,
+        engine.language
+    );
+
+    if (!unlimitedSelectQuery) {
+        return query;
+    }
+
+    // Show a warning modal to let user confirm what they are doing
+    return new Promise<string>((resolve, reject) => {
+        sendConfirm({
+            header: 'Your SELECT query is unbounded',
+            message: (
+                <Content>
+                    <div>
+                        The following SELECT statement have no limit. Please
+                        make sure you intend to get all the rows returned.
+                    </div>
+                    <div>
+                        <i>
+                            Tip: to avoid seeing this message, add a LIMIT in
+                            the query or set it on left of the run button.
```

Comment:
not good at English grammar, but feel like it should be `on the left` :) 

or maybe we can say `or pick a limit from the Limit dropdown list`